File: test/config.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { configSchema } from "../src/config.js";
import { iterationHasEnded, daysBetween, hoursBetween, iterationEnd } from "../src/util/dates.js";

test("config applies field-name defaults", () => {
  const cfg = configSchema.parse({ project: { owner: "acme", number: 3 }, features: {} });
  assert.equal(cfg.project.type, "org");
  assert.equal(cfg.fields.status, "Status");
  assert.deepEqual(cfg.doneStatuses, ["Done"]);
  assert.equal(cfg.features.rollover.enabled, false);
});

test("config rejects a missing project number", () => {
  const res = configSchema.safeParse({ project: { owner: "acme" }, features: {} });
  assert.equal(res.success, false);
});

test("config rejects a postTo target with neither issue nor createIssueTitle", () => {
  const res = configSchema.safeParse({
    project: { owner: "acme", number: 1 },
    features: { digest: { enabled: true, postTo: {} } },
  });
  assert.equal(res.success, false);
});

test("iterationHasEnded is true once the window has elapsed", () => {
  assert.equal(iterationHasEnded("2026-06-01", 14, new Date("2026-06-16T00:00:00Z")), true);
  assert.equal(iterationHasEnded("2026-06-01", 14, new Date("2026-06-10T00:00:00Z")), false);
});

test("daysBetween counts fractional days", () => {
  assert.equal(daysBetween(new Date("2026-07-01T00:00:00Z"), new Date("2026-07-04T00:00:00Z")), 3);
});

test("hoursBetween counts fractional hours", () => {
  assert.equal(hoursBetween(new Date("2026-07-01T00:00:00Z"), new Date("2026-07-01T06:30:00Z")), 6.5);
});

test("iterationEnd is start + duration days (exclusive)", () => {
  assert.equal(iterationEnd("2026-06-01", 14).toISOString(), "2026-06-15T00:00:00.000Z");
});

File: ruby/test/config_test.rb
# frozen_string_literal: true

require_relative "test_helper"

class ConfigTest < Minitest::Test
  def test_applies_defaults
    cfg = Boardly::Config.new(project: { owner: "acme", number: 3 }, features: {})
    assert_equal "org", cfg.project[:type]
    assert_equal "Status", cfg.fields[:status]
    assert_equal ["Done"], cfg.done_statuses
    assert_equal false, cfg.features[:rollover][:enabled]
  end

  def test_rejects_missing_project_number
    err = assert_raises(Boardly::ConfigError) { Boardly::Config.new(project: { owner: "acme" }, features: {}) }
    assert_match(/project.number/, err.message)
  end

  def test_rejects_post_to_without_issue_or_title
    err = assert_raises(Boardly::ConfigError) do
      Boardly::Config.new(project: { owner: "acme", number: 1 }, features: { digest: { enabled: true, post_to: {} } })
    end
    assert_match(/postTo: requires/, err.message)
  end

  def test_notifications_slack_defaults
    cfg = Boardly::Config.new(project: { owner: "acme", number: 1 }, features: {}, notifications: { slack: { enabled: true } })
    assert_equal "SLACK_WEBHOOK_URL", cfg.notifications[:slack][:webhook_env]
  end
end
